nuxeo-runtime/nuxeo-runtime-stream/src/main/java/org/nuxeo/runtime/avro/AvroServiceImpl.java

```diff
@@ -0,0 +1,151 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.message.SchemaStore;
+import org.nuxeo.runtime.RuntimeServiceException;
+
+/**
+ * @since 10.2
+ */
+public class AvroServiceImpl
+        extends SchemaStore.Cache
+        implements AvroService {
+
+    private static final AvroMapper<Object, Object> NULL = new AvroMapper<Object, Object>(null) {
+
+        @Override
+        public Object fromAvro(Schema schema, Object input) {
+            return null;
+        }
+
+        @Override
+        public GenericRecord toAvro(Schema schema, Object input) {
+            return null;
+        }
+
+    };
+
+    protected final Map<Class<?>, Class<AvroSchemaFactory<?>>> factories;
+
+    protected final List<AvroReplacementDescriptor> replacements;
+
+    protected Map<Class<?>, AvroMapper<?, ?>> mappers;
+
+    public AvroServiceImpl(Collection<AvroReplacementDescriptor> replacements,
+            Map<Class<?>, Class<AvroSchemaFactory<?>>> factories) {
+        this.replacements = new ArrayList<>(replacements);
+        this.factories = new HashMap<>(factories);
+        this.replacements.sort(null);
+        // assert at creation that factories are valid
+        createContext();
+    }
+
+    @Override
+    public <D> Schema createSchema(D input) {
+        return createContext().createSchema(input);
+    }
+
+    @Override
+    public String decodeName(String input) {
+        String output = input;
+        ListIterator<AvroReplacementDescriptor> it = replacements.listIterator(replacements.size());
+        while (it.hasPrevious()) {
+            AvroReplacementDescriptor replacement = it.previous();
+            output = output.replaceAll(replacement.getReplacement(), replacement.getForbidden());
+        }
+        return output;
+    }
+
+    @Override
+    public String encodeName(String input) {
+        String output = input;
+        for (AvroReplacementDescriptor replacement : replacements) {
+            output = output.replaceAll(replacement.getForbidden(), replacement.getReplacement());
+        }
+        return output;
+    }
+
+    @Override
+    public Schema findByName(String name) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <D, M> D fromAvro(Schema schema, Class<D> clazz, M input) {
+        return (D) ((AvroMapper<D, M>) getMapper(clazz)).fromAvro(schema, input);
+    }
+
+    public void setMappers(Map<Class<?>, AvroMapper<?, ?>> mappers) {
+        this.mappers = new HashMap<>(mappers);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <D, M> M toAvro(Schema schema, D input) {
+        return ((AvroMapper<D, M>) getMapper(input.getClass())).toAvro(schema, input);
```

Comment:
What about making `getMapper` generic and move the cast into it ?

Reply:
done